File: common/java/common-data-storage/src/main/java/com/welab/wefe/common/data/storage/repo/impl/ClickhouseStorage.java
/*
 * Copyright 2021 Tianmian Tech. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.welab.wefe.common.data.storage.repo.impl;

import com.welab.wefe.common.data.storage.model.DataItemModel;
import com.welab.wefe.common.data.storage.model.PageInputModel;
import com.welab.wefe.common.data.storage.model.PageOutputModel;
import com.welab.wefe.common.data.storage.repo.AbstractStorage;
import net.razorvine.pickle.Pickler;
import net.razorvine.pickle.Unpickler;
import org.springframework.stereotype.Component;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/**
 * @author yuxin.zhang
 */
@Component
public class ClickhouseStorage extends AbstractStorage {

    @Override
    public void put(String dbName, String tbName, DataItemModel model) throws Exception {
        Connection conn = null;
        PreparedStatement statement = null;
        try {
            checkTB(dbName, tbName);
            conn = getConnection();
            String sql = "INSERT INTO " + formatTableName(dbName, tbName) + " (eventDate,k,v,id) values(?,?,?,?)";
            statement = conn.prepareStatement(sql);
            Pickler pickler = new Pickler();
            statement.setDate(1, model.getEventDate());
            byte[] key = model.getK() instanceof byte[] ? (byte[]) model.getK() : pickler.dumps(model.getK());
            byte[] value = model.getV() instanceof byte[] ? (byte[]) model.getV() : pickler.dumps(model.getV());
            statement.setBytes(2, key);
            statement.setBytes(3, value);
            statement.setString(4, UUID.randomUUID().toString());
            statement.executeQuery();
        } finally {
            close(statement, conn);
        }
    }

    @Override
    public <K, V> void putAll(String dbName, String tbName, List<DataItemModel<K, V>> list) throws Exception {
        Connection conn = null;
        PreparedStatement statement = null;
        try {
            checkTB(dbName, tbName);
            conn = getConnection();
            conn.setAutoCommit(false);
            String sql = "INSERT INTO " + formatTableName(dbName, tbName) + " (eventDate,k,v,id) values(?,?,?,?)";
            statement = conn.prepareStatement(sql);
            Pickler pickler = new Pickler();
            for (DataItemModel<K, V> item : list) {
                statement.setDate(1, item.getEventDate());
                byte[] key = item.getK() instanceof byte[] ? (byte[]) item.getK() : pickler.dumps(item.getK());
                byte[] value = item.getV() instanceof byte[] ? (byte[]) item.getV() : pickler.dumps(item.getV());
                statement.setBytes(2, key);
                statement.setBytes(3, value);
                statement.setString(4, UUID.randomUUID().toString());
                statement.addBatch();
            }
            statement.executeBatch();
            conn.commit();
        } finally {
            close(statement, conn);
        }
    }

    @Override
    public DataItemModel get(String dbName, String tbName, String key) throws Exception {
        DataItemModel model = null;
        Connection conn = null;
        PreparedStatement statement = null;
        ResultSet rs = null;
        try {
            conn = getConnection();
            String sql = "select * from " + formatTableName(dbName, tbName) + " where k=? order by id DESC";
            statement = conn.prepareStatement(sql);
            Pickler pickler = new Pickler();
            Unpickler unpickler = new Unpickler();
            statement.setBytes(1, pickler.dumps(key));
            rs = statement.executeQuery();
            if (rs.next()) {
                model = new DataItemModel();
                Date eventDate = rs.getDate(1);
                model.setEventDate(eventDate);
                model.setK(unpickler.loads(rs.getBytes(2)));
                model.setV(unpickler.loads(rs.getBytes(3)));
            }
        } finally {
            close(rs, statement, conn);
        }

        return model;
    }

    @Override
    public List<DataItemModel> collect(String dbName, String tbName) throws Exception {
        List<DataItemModel> list;
        Connection conn = null;
        PreparedStatement statement = null;
        ResultSet rs = null;
        try {
            conn = getConnection();
            String sql = "select * from " + formatTableName(dbName, tbName) + "  order by id DESC";
            statement = conn.prepareStatement(sql);
            rs = statement.executeQuery();
            list = new ArrayList();
            Unpickler unpickler = new Unpickler();
            while (rs.next()) {
                DataItemModel model = new DataItemModel();
                Date eventDate = rs.getDate(1);
                model.setEventDate(eventDate);
                model.setK(unpickler.loads(rs.getBytes(2)));
                model.setV(unpickler.loads(rs.getBytes(3)));
                list.add(model);
            }
        } finally {
            close(rs, statement, conn);
        }

        return list;
    }


    @Override
    public List<DataItemModel<byte[], byte[]>> collectBytes(String dbName, String tbName) throws Exception {
        List<DataItemModel<byte[], byte[]>> list;
        Connection conn = null;
        PreparedStatement statement = null;
        ResultSet rs = null;
        try {
            conn = getConnection();
            String sql = "select * from " + formatTableName(dbName, tbName);
            statement = conn.prepareStatement(sql);
            rs = statement.executeQuery();
            list = new ArrayList();
            while (rs.next()) {
                DataItemModel<byte[], byte[]> model = new DataItemModel();
                Date eventDate = rs.getDate(1);
                model.setEventDate(eventDate);
                model.setK(rs.getBytes(2));
                model.setV(rs.getBytes(3));
                list.add(model);
            }
        } finally {
            close(rs, statement, conn);
        }

        return list;
    }

    @Override
    public void delete(String dbName, String tbName, String key) throws Exception {
        Connection conn = null;
        PreparedStatement statement = null;
        try {
            conn = getConnection();
            String sql = "alter table " + formatTableName(dbName, tbName) + " delete where k=?";
            statement = conn.prepareStatement(sql);
            Pickler pickler = new Pickler();
            statement.setBytes(1, pickler.dumps(key));
            statement.executeQuery();
        } finally {
            close(statement, conn);
        }
    }

    @Override
    public int count(String dbName, String tbName) throws Exception {
        int result = 0;
        Connection conn = null;
        PreparedStatement statement = null;
        ResultSet rs = null;
        try {
            conn = getConnection();
            String sql = "select count(*) from " + formatTableName(dbName, tbName);
            statement = conn.prepareStatement(sql);
            rs = statement.executeQuery();
            if (rs.next()) {
                result = rs.getInt(1);
            }
        } finally {
            close(rs, statement, conn);
        }
        return result;
    }

    @Override
    public List<DataItemModel> take(String dbName, String tbName, int size) throws Exception {
        if (size < 0) {
            size = 0;
        }
        List<DataItemModel> result = new ArrayList<>();
        List<DataItemModel> list = collect(dbName, tbName);
        for (int i = 0; i < list.size(); i++) {
            if (i == size) {
                break;
            }
            result.add(list.get(i));
        }
        return result;
    }

    @Override
    public PageOutputModel getPage(String dbName, String tbName, PageInputModel pageInputModel) throws Exception {
        PageOutputModel pageOutputModel = new PageOutputModel();
        int total = count(dbName, tbName);
        pageOutputModel.setTotal(total);

        List<DataItemModel> list;
        Connection conn = null;
        PreparedStatement statement = null;
        ResultSet rs = null;
        try {
            conn = getConnection();
            String sql = "select * from " + formatTableName(dbName, tbName) + "  order by id DESC limit ?,?";
            statement = conn.prepareStatement(sql);
            int start = pageInputModel.getPageNum() == 0 ? 0 : pageInputModel.getPageNum() * pageInputModel.getPageSize();
            statement.setInt(1, start);
            statement.setInt(2, pageInputModel.getPageSize());
            rs = statement.executeQuery();
            list = new ArrayList();
            Unpickler unpickler = new Unpickler();
            while (rs.next()) {
                DataItemModel model = new DataItemModel();
                Date eventDate = rs.getDate(1);
                model.setEventDate(eventDate);
                model.setK(unpickler.loads(rs.getBytes(2)));
                model.setV(unpickler.loads(rs.getBytes(3)));
                list.add(model);
            }
            pageOutputModel.setCurrentNum(pageInputModel.getPageNum());
            pageOutputModel.setTotalPages((total + pageInputModel.getPageSize() - 1) / pageInputModel.getPageSize());
            pageOutputModel.setData(list);
        } finally {
            close(rs, statement, conn);
        }

        return pageOutputModel;
    }


    @Override
    public PageOutputModel<byte[], byte[]> getPageBytes(String dbName, String tbName, PageInputModel pageInputModel) throws Exception {
        PageOutputModel<byte[], byte[]> pageOutputModel = new PageOutputModel();
        int total = count(dbName, tbName);
        pageOutputModel.setTotal(total);

        List<DataItemModel<byte[], byte[]>> list;
        Connection conn = null;
        PreparedStatement statement = null;
        ResultSet rs = null;
        try {
            conn = getConnection();
            String sql = "select * from " + formatTableName(dbName, tbName) + "  order by id DESC limit ?,?";
            statement = conn.prepareStatement(sql);
            int start = pageInputModel.getPageNum() == 0 ? 0 : pageInputModel.getPageNum() * pageInputModel.getPageSize();
            statement.setInt(1, start);
            statement.setInt(2, pageInputModel.getPageSize());
            rs = statement.executeQuery();
            list = new ArrayList();
            while (rs.next()) {
                DataItemModel model = new DataItemModel();
                Date eventDate = rs.getDate(1);
                model.setEventDate(eventDate);
                model.setK(rs.getBytes(2));
                model.setV(rs.getBytes(3));
                list.add(model);
            }
            pageOutputModel.setCurrentNum(pageInputModel.getPageNum());
            pageOutputModel.setTotalPages((total + pageInputModel.getPageSize() - 1) / pageInputModel.getPageSize());
            pageOutputModel.setData(list);
        } finally {
            close(rs, statement, conn);
        }
        return pageOutputModel;
    }

    @Override
    public void dropTB(String dbName, String tbName) throws Exception {
        Connection conn = null;
        PreparedStatement statement = null;
        try {
            conn = getConnection();
            String sql = "DROP TABLE " + formatTableName(dbName, tbName);
            statement = conn.prepareStatement(sql);
            statement.execute();
        } finally {
            close(statement, conn);
        }
    }

    @Override
    public void dropDB(String dbName) throws Exception {
        Connection conn = null;
        PreparedStatement statement = null;
        try {
            conn = getConnection();
            String sql = "DROP DATABASE " + String.format("`%s`", dbName);
            statement = conn.prepareStatement(sql);
            statement.execute();
        } finally {
            close(statement, conn);
        }
    }

    @Override
    public int getAddBatchSize(int columnCount) {
        return 150000 / columnCount;
    }

    @Override
    public int getCountByByteSize(String dbName, String tbName, long byteSize) throws Exception {
        Connection conn = null;
        PreparedStatement statement = null;
        ResultSet rs = null;

        try {
            conn = getConnection();
            String sql = "select * from " + formatTableName(dbName, tbName) + " limit 0,1";
            statement = conn.prepareStatement(sql);
            rs = statement.executeQuery();
            if (rs.next()) {
                byte[] key = rs.getBytes(2);
                byte[] value = rs.getBytes(3);

                // The size of a single piece of data
                int unitSize = key.length + value.length;
                if (unitSize != 0) {
                    //Calculate the optimal page size
                    int count = BigDecimal.valueOf(byteSize)
                            .divide(BigDecimal.valueOf(unitSize), 0, RoundingMode.HALF_UP)
                            .intValue();
                    return count > 0 ? count : super.getCountByByteSize(dbName, tbName, byteSize);
                }
            }
        } finally {
            close(rs, statement, conn);
        }
        return super.getCountByByteSize(dbName, tbName, byteSize);
    }

    @Override
    public boolean isExists(String dbName, String tbName) throws SQLException {
        Connection conn = null;
        PreparedStatement statement = null;
        try {
            conn = getConnection();
            String sql = String.format("SELECT  count(*) from system.parts p  where active  and database = '%s' and table ='%s'", dbName, tbName);
            statement = conn.prepareStatement(sql);
            ResultSet rs = statement.executeQuery();
            if(rs.next()) {
                return rs.getInt(1) > 0;
            }

            return false;
//            System.out.println();
        } finally {
            close(statement, conn);
        }
//        return false;
    }
}

File: common/python/storage/impl/clickhouse_storage.py
# Copyright 2021 Tianmian Tech. All Rights Reserved.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import datetime
import uuid
from collections.abc import Iterable
from concurrent.futures import ThreadPoolExecutor, as_completed
from fnmatch import fnmatch

from clickhouse_driver import Client

from common.python.common import consts
from common.python.common.consts import NAMESPACE
from common.python.storage.storage import Storage
from common.python.utils import conf_utils
from common.python.utils.core_utils import deserialize

FORCE_SERI = False


def set_force_serialize(in_seri):
    return in_seri if not FORCE_SERI else FORCE_SERI


def get_db_conn(database='default'):
    env_ck_host = conf_utils.get_env_config(consts.ENV_CONF_KEY_CK_HOST)
    env_ck_port = conf_utils.get_env_config(consts.ENV_CONF_KEY_CK_PORT)

    ck_host = conf_utils.get_comm_config(consts.COMM_CONF_KEY_CK_HOST)
    ck_user = conf_utils.get_comm_config(consts.COMM_CONF_KEY_CK_USER)
    ck_pwd = conf_utils.get_comm_config(consts.COMM_CONF_KEY_CK_PWD)
    ck_port = conf_utils.get_comm_config(consts.COMM_CONF_KEY_CK_PORT)
    return Client(host=env_ck_host or ck_host, user=ck_user, password=ck_pwd, database=database,
                  port=env_ck_port or ck_port)


class ClickHouseStorage(Storage):

    def __init__(self, _type, namespace, name, partitions=1, in_place_computing=False):
        self._type = _type
        self._namespace = namespace
        self._name = name
        self.table_name = f"`{namespace}`.`{name}`"
        self._partitions = partitions
        self.schema = {}
        self._in_place_computing = in_place_computing

        # self.init_db()
        self.init_tb()

    def init_db(self):
        sql = f"CREATE DATABASE IF NOT EXISTS `{self._namespace}`"
        try:
            client = self.get_conn()
            client.execute(sql)
        finally:
            client.disconnect()

    def init_tb(self):
        sql = f"CREATE TABLE IF NOT EXISTS {self.table_name}(`eventDate` Date, `k` String, `v` String, `id` String)" \
              f"ENGINE = MergeTree() PARTITION BY toDate(eventDate) ORDER BY (id) SETTINGS index_granularity = 8192"
        try:
            client = self.get_conn()
            client.execute(sql)
        finally:
            client.disconnect()

    def get_conn(self):
        return get_db_conn()

    def put(self, k, v, use_serialize=True):
        use_serialize = set_force_serialize(use_serialize)
        k_bytes, v_bytes = self.kv_to_bytes(k=k, v=v, use_serialize=use_serialize)
        sql = f"INSERT INTO {self.table_name} (eventDate,k,v,id) values"
        try:
            client = self.get_conn()
            result = client.execute(sql,
                                    [[datetime.datetime.now(), k_bytes, v_bytes, str(uuid.uuid1())]])
        finally:
            client.disconnect()
        return result

    def ck_execute(self, sql, data):
        client = self.get_conn()
        client.execute(sql, data)
        client.disconnect()

    def put_all(self, kv_list: Iterable, use_serialize=True, chunk_size=100000):
        use_serialize = set_force_serialize(use_serialize)
        sql = f"INSERT INTO {self.table_name} (eventDate,k,v,id) values"
        write_batch = 10000
        max_workers = None
        try:
            with ThreadPoolExecutor(max_workers=max_workers) as t:
                data = []
                all_task = []
                for k, v in kv_list:
                    k_bytes, v_bytes = self.kv_to_bytes(k=k, v=v, use_serialize=use_serialize)
                    data.append([datetime.datetime.now(), k_bytes, v_bytes, str(uuid.uuid1())])
                    if len(data) == write_batch:
                        all_task.append(t.submit(self.ck_execute, sql, data))
                        data = []
                if len(data) > 0:
                    all_task.append(t.submit(self.ck_execute, sql, data))

                for future in as_completed(all_task):
                    future.result()
        finally:
            pass

    def put_if_absent(self, k, v, use_serialize=True):
        use_serialize = set_force_serialize(use_serialize)
        exist_v = self.get(k, use_serialize=use_serialize)
        if not exist_v:
            return self.put(k, v, use_serialize)

    def get(self, k, use_serialize=True, maybe_large_value=False):
        use_serialize = set_force_serialize(use_serialize)
        k_hex = self.kv_to_bytes(k=k, use_serialize=use_serialize).hex().upper()
        sql = f"SELECT v FROM {self.table_name} WHERE hex(k)=%(a)s ORDER BY id DESC LIMIT 1"
        client = self.get_conn()
        try:
            result = client.execute(sql, {"a": k_hex})
            data = None if len(result) == 0 else result[0][0]
            return None if data is None else (
                deserialize(data) if use_serialize else data)
        finally:
            client.disconnect()

    def collect(self, min_chunk_size=0, use_serialize=True, partition=None) -> list:
        use_serialize = set_force_serialize(use_serialize)
        sql = f"SELECT k,v FROM {self.table_name} ORDER BY id DESC"
        try:
            client = self.get_conn()
            data = client.execute_iter(sql)
            for item in data:
                key, value = item[0], item[1]
                if use_serialize:
                    yield deserialize(key), deserialize(value)
                else:
                    yield key, value
                    # yield bytes_to_string(key), value
        finally:
            client.disconnect()

    def delete(self, k, use_serialize=True):
        use_serialize = set_force_serialize(use_serialize)
        try:
            client = self.get_conn()
            value = self.get(k)
            k_hex = self.kv_to_bytes(k=k, use_serialize=use_serialize).hex().upper()
            sql = f"ALTER TABLE {self.table_name} DELETE WHERE hex(k)=%(a)s"
            client.execute(sql, {"a": k_hex})
        finally:
            client.disconnect()
        return value

    def destroy(self):
        try:
            client = self.get_conn()
            sql = f"DROP TABLE {self.table_name}"
            print(f"destory: {sql}")
            client.execute(sql)
        finally:
            client.disconnect()

        _table_key = ".".join([self._type, self._namespace, self._name])
        from common.python.p_session.base_impl.db_runtime import DBRuntime
        DBRuntime.get_instance().meta_table.delete(_table_key)

    def count(self):
        sql = f"SELECT count(*) FROM {self.table_name}"
        try:
            client = self.get_conn()
            result = client.execute(sql)
            result = result[0][0]
        finally:
            client.disconnect()
        return result

    def take(self, n=1, keysOnly=False, use_serialize=True):
        use_serialize = set_force_serialize(use_serialize)

        if n <= 0:
            n = 1
        it = self.collect(use_serialize=use_serialize)
        rtn = list()
        i = 0
        for item in it:
            if keysOnly:
                rtn.append(item[0])
            else:
                rtn.append(item)
            i += 1
            if i == n:
                break
        return rtn

    def first(self, keysOnly=False, use_serialize=True):
        use_serialize = set_force_serialize(use_serialize)

        resp = self.take(1, keysOnly=keysOnly, use_serialize=use_serialize)
        if resp:
            return resp[0]
        else:
            return None

    def save_as(self, name, namespace, partition=None, use_serialize=True, **kwargs):
        use_serialize = set_force_serialize(use_serialize)

        if partition is None:
            partition = self._partitions

        persistent = kwargs["persistent"]
        persistent_engine = kwargs["persistent_engine"]
        dsource = kwargs["dsource"]

        dsource.set_gc_disable()

        from common.python.p_session.base_impl.db_runtime import DBRuntime
        dup = DBRuntime.get_instance().table(name, namespace, partition,
                                             persistent=persistent, persistent_engine=persistent_engine)

        new_table_name = f"`{namespace}`.`{name}`"
        try:
            client = self.get_conn()
            sql = f"INSERT INTO {new_table_name} SELECT * FROM {self.table_name} ORDER BY k DESC"
            client.execute(sql)
        finally:
            client.disconnect()

        dsource.set_gc_enable()
        return dup


def del_all_database():
    """
    Delete all database for test

    Returns
    -------

    """
    storage = ClickHouseStorage(_type=None, namespace="wefe", name="test")
    conn = storage.get_conn()
    sql = "show databases"

    exclude = ["wefe", "system", "default", "wefe_yuxin"]
    result = conn.execute(sql)
    print(result)
    for item in result:
        db_name = item[0]
        if db_name not in exclude:
            drop_sql = f"drop database if exists `{db_name}`"
            conn.execute(drop_sql)

    conn.disconnect()


def clean_up_tables(name_pattern, namespace=None):
    client = None
    try:
        if not namespace:
            namespace = NAMESPACE.PROCESS

        if not name_pattern:
            return

        client = get_db_conn(namespace)
        sql = "show tables"
        result = client.execute(sql)
        for item in result:
            db_name = item[0]
            if fnmatch(db_name, name_pattern):
                drop_sql = f"drop table `{namespace}`.`{db_name}`"
                client.execute(drop_sql)
    finally:
        if client:
            client.disconnect()


if __name__ == '__main__':
    storage = ClickHouseStorage(_type=None, namespace="wefe_data", name="111111111111111")
    # clean_up_tables("442a9be30a66423c9798e6ae18a152a0")
    storage.put_all([(1, 1), (2, 2)])
    # print(list(storage.collect()))
    # del_all_database()

    # storage.save_as(name="yuxin_auto_create_test",namespace="wefe")
